### PR TITLE
Fix for deserializing imported ordered dictionary

### LINKED ISSUE #2861: Deserialized Ordered Dictionary via Import-Clixml is not numerically indexable.
Steps to reproduce
------------------

```
# create ordered dictionary via [ordered] syntactic sugar
$a = [ordered]@{}
$a['Larry'] = 'Poik!'
$a['Curly'] = 'Nyuk!'
$a['Moe'] = 'Wise guy!'

# get 2nd item from ordered hash
$a[1]
# returns: Nyuk!

# Persist to disk
$a | Export-Clixml ./tempObject.xml

# Deserialize
$b = Import-Clixml ./tempObject.xml

# get 2nd item from rehydrated ordered hash
$b[1]
```

Expected behavior
-----------------

```
# returns: Nyuk!
```

Actual behavior
---------------

```
# returns: $null

$a | Get-Member
   TypeName: System.Collections.Specialized.OrderedDictionary
...

$b | Get-Member
   TypeName: Deserialized.System.Collections.Specialized.OrderedDictionary
...
```
Deserialized object does not have the full implementation of the original OrderedDictionary object.

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
SerializationVersion           1.1.0.1
PSEdition                      Core
BuildVersion                   3.0.0.0
PSRemotingProtocolVersion      2.3
WSManStackVersion              3.0
PSVersion                      6.0.0-alpha
GitCommitId                    v6.0.0-alpha.13-42-gf38d09dc6d7bf208cd78d86e1...
CLRVersion
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
```

### PR DESCRIPTION
**Pull request by davidBar-On**
# PR Summary

Suggested fix for issue #2861 - exported ordered dictionary is now imported via `Import-Clixml` as `OrderedDictionary` and not as `Hashtable`.  This is done by checking if the imported dictionary type names include type of `OrderedDictionary`.

## PR Context

Fix #2861.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/serialization.cs [hunk 1]
```diff
@@ -3992,14 +3993,27 @@ private object ReadListContainer(ContainerType ct)
...
+            // Find whether original directory was odered or not
+            bool isOrdered = false;
+            if (InternalTypeNames.Count > 0)
+            {
+                string originalDictTypeName = Deserializer.MaskDeserializationPrefix(InternalTypeNames[0]);
+                if (originalDictTypeName == tableOrdered.GetType().FullName)
+                {
+                    isOrdered = true;
+                }
+            }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/serialization.cs:4015]: There is a typo in comment - "odered".

Also we could pack the code in one line initialization:
```suggestion
            // Find whether original directory was ordered or not
            bool isOrdered = InternalTypeNames.Count > 0 && (Deserializer.MaskDeserializationPrefix(InternalTypeNames[0]) == tableOrdered.GetType().FullName));
```

File: src/System.Management.Automation/engine/serialization.cs [hunk 2]
```diff
@@ -3992,14 +3993,27 @@ private object ReadListContainer(ContainerType ct)
...
-        private object ReadDictionary(ContainerType ct)
+        private object ReadDictionary(ContainerType ct, ConsolidatedString InternalTypeNames)
         {
             Dbg.Assert(ct == ContainerType.Dictionary, "Unrecognized ContainerType enum");
 
             // We assume the hash table is a PowerShell hash table and hence uses
             // a case insensitive string comparer.  If we discover a key collision,
             // we'll revert back to the default comparer.
             Hashtable table = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+            OrderedDictionary tableOrdered = new OrderedDictionary(StringComparer.CurrentCultureIgnoreCase);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/serialization.cs:4004]: Why do we name the variable as table if it is a dictionary? Maybe `orderedDictionary`.

File: src/System.Management.Automation/engine/serialization.cs [hunk 3]
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
 
                     object value = ReadOneObject();
 
-                    // On the first collision, copy the hash table to one that uses the default comparer.
-                    if (table.ContainsKey(key) && (keyClashFoundIteration == 0))
+                    if (!isOrdered)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/serialization.cs:4056]: Why don't swap and start with `if (isOrdered)`?

File: src/System.Management.Automation/engine/serialization.cs [hunk 4]
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
...
                     {
-                        keyClashFoundIteration++;
-                        Hashtable newHashTable = new Hashtable();
-                        foreach (DictionaryEntry entry in table)
+                        // On the first collision, copy the hash table to one that uses the default comparer.
+                        if (table.ContainsKey(key) && (keyClashFoundIteration == 0))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/serialization.cs:4059]:
```suggestion
                        if (keyClashFoundIteration == 0 && table.ContainsKey(key))
```

Below too.

File: src/System.Management.Automation/engine/serialization.cs [hunk 5]
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
...
                         {
-                            newHashTable.Add(entry.Key, entry.Value);
+                            keyClashFoundIteration++;
+                            Hashtable newHashTable = new Hashtable();
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/serialization.cs:4062]:
```suggestion
                            Hashtable newHashTable = new Hashtable(table.Count);
```

Below too.

File: src/System.Management.Automation/engine/serialization.cs [hunk 6]
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
...
-                        keyClashFoundIteration++;
-                        IEqualityComparer equalityComparer = new ReferenceEqualityComparer();
-                        Hashtable newHashTable = new Hashtable(equalityComparer);
-                        foreach (DictionaryEntry entry in table)
+                        // On the first collision, copy the hash table to one that uses the default comparer.
+                        if (tableOrdered.Contains(key) && (keyClashFoundIteration == 0))
                         {
-                            newHashTable.Add(entry.Key, entry.Value);
+                            keyClashFoundIteration++;
+                            OrderedDictionary newOrderedTable = new OrderedDictionary();
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/serialization.cs:4092]:
```suggestion
                            OrderedDictionary newOrderedTable = new OrderedDictionary(tableOrdered.Count);
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 7]
```diff
@@ -52,6 +52,31 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
     }
 
+	It "Import from exported non-ordered dictionary object should create non-ordered dictionary object" {
+		$dict = @{}
+		$dict['Larry'] = 'Poik!'
+		$dict['Curly'] = 'Nyuk!'
+		$dict['Moe'] = 'Wise guy!'
+		$dict | Export-Clixml $testfile
+		$results = Import-Clixml $testfile
+		$results.GetType().Name | Should -Be "Hashtable"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:62]: Please use `BeExactly` for strings:
```suggestion
		$results.GetType().Name | Should -BeExactly "Hashtable"
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 8]
```diff
@@ -52,6 +52,31 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
...
+    }
+
+	It "Import from exported ordered dictionary object should create ordered dictionary object" {
+		$dict = [ordered]@{}
+		$dict['Larry'] = 'Poik!'
+		$dict['Curly'] = 'Nyuk!'
+		$dict['Moe'] = 'Wise guy!'
+		$dict | Export-Clixml $testfile
+		$results = Import-Clixml $testfile
+		$results.GetType().Name | Should -Be "OrderedDictionary"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:73]:
```suggestion
		$results.GetType().Name | Should -BeExactly "OrderedDictionary"
```

File: src/System.Management.Automation/engine/serialization.cs (hunk 1 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: Changed.

File: src/System.Management.Automation/engine/serialization.cs (hunk 2 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: Name changed.

File: src/System.Management.Automation/engine/serialization.cs (hunk 3 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: Swapped.

File: src/System.Management.Automation/engine/serialization.cs (hunk 4 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: This was a copy from the original code - changed in all 4 places (not/ordered x iteration=0/1).

File: src/System.Management.Automation/engine/serialization.cs (hunk 5 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: Changed in both cases (again, was in the original code).

File: src/System.Management.Automation/engine/serialization.cs (hunk 6 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: Changed in both cases.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 (hunk 7 above)
Comment (davidBar-On) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Changed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 (hunk 8 above)
Comment (davidBar-On) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Changed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 9]
```diff
@@ -59,8 +59,8 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
 		$dict['Moe'] = 'Wise guy!'
 		$dict | Export-Clixml $testfile
 		$results = Import-Clixml $testfile
-		$results.GetType().Name | Should -Be "Hashtable"
-		$results.Count | Should -Be 3
+		$results.GetType().Name | Should -BeExactly  "Hashtable"
+		$results.Count | Should -BeExactly 3
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:63]: It is not string.
```suggestion
		$results.Count | Should -Be 3
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 [hunk 10]
```diff
@@ -70,11 +70,11 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
 		$dict['Moe'] = 'Wise guy!'
 		$dict | Export-Clixml $testfile
 		$results = Import-Clixml $testfile
-		$results.GetType().Name | Should -Be "OrderedDictionary"
-		$results.Count | Should -Be 3
-		$results[0] | Should -Be $dict[0]
-		$results[1] | Should -Be $dict[1]
-		$results[2] | Should -Be $dict[2]
+		$results.GetType().Name | Should -BeExactly  "OrderedDictionary"
+		$results.Count | Should -BeExactly  3
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:74]:
```suggestion
		$results.Count | Should -Be  3
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 (hunk 9 above)
Comment (davidBar-On) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1 (hunk 10 above)
Comment (davidBar-On) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1]: Done

**Review by iSazonov — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@PaulHigin can you please take a look at this PR and see the change to `serailization.cs` is the right fix?

**Review by PaulHigin — CHANGES_REQUESTED** [review]
These changes look right to me, except that instead of having if/else blocks, I feel it would be cleaner if you created a helper class that encapsulates the dictionary types (Hashtable or OrderedDictionary) and handles adding key/value pairs and re-creating with case sensitive comparer as needed.  Something like:

```csharp
private class PSDictionary
{
    private Hashtable table;
    private OrderedDictionary orderedDictionary;

    public PSDictionary(bool isOrdered) { ... }

    public Add(object key, object value) { ... }

    public object DictionaryObject { get { return _isOrdered ? orderedDictionary : table; } }
}
```

Also, I think it is fine to compare the type name check to a hard coded constant string: `System.Collections.Specialized.OrderedDictionary`, instead of getting the full type name from an instance of the object.

**Comment by davidBar-On** [comment]
> ... instead of having if/else blocks, I feel it would be cleaner if you created a helper class .....

Done

> Also, I think it is fine to compare the type name check to a hard coded constant string: `System.Collections.Specialized.OrderedDictionary`, instead of getting the full type name from an instance of the object.

Did not make the this change, as it seems to me that it safer not to assume the type name.  Note that with the other changes `FullName` is now for the class `typeOf` and not for a variable.

File: src/System.Management.Automation/engine/serialization.cs [hunk 11]
```diff
@@ -3986,19 +3987,104 @@ private object ReadListContainer(ContainerType ct)
...
+                    table = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+                }
+            }
+
+            public object DictionaryObject { get { return _isOrdered ? orderedDictionary : table; } }
+
+            public void Add(object key, object value) {
+                // On the first collision, copy the hash table to one that uses the default comparer.
+                _containsKey = _isOrdered ? orderedDictionary.Contains(key) : table.ContainsKey(key);
+                if (_keyClashFoundIteration == 0 && _containsKey)
```
Comment (TravisEz13) [inline src/System.Management.Automation/engine/serialization.cs:4017]: Can you add a test for a key clash?

Comment (TravisEz13) [reply src/System.Management.Automation/engine/serialization.cs:4017]: Both kinds

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
see comments

File: src/System.Management.Automation/engine/serialization.cs (hunk 11 above)
Comment (davidBar-On) [reply src/System.Management.Automation/engine/serialization.cs]: Added test cases.  See the general discussion comments.

**Comment by davidBar-On** [comment]
> Can you add a test for a key clash? Both kinds

Added test cases for the keys clash.  However, for most of the cases I created an "exported" XML manually, as I did not find a way to create hashtable (ordered or not) that includes duplicate keys, or an ordered hashtable that allows case-sensitive keys.  The only added test without using XML is for non-ordered case-sensitive hashtable using `New-Object hashtable` instead of `@{}`.

In the added tests I didn't find a way to check which "Comparer" is set for the imported table.  Therefore the tests only check the number of elements in the imported table.

Also note that I rewrote the change using the `IDictionary` interface to simplify the implementation.

File: src/System.Management.Automation/engine/serialization.cs [hunk 12]
```diff
@@ -3986,19 +3987,81 @@ private object ReadListContainer(ContainerType ct)
...
+        private class PSDictionary
+        {
+            private IDictionary dict;
+            private readonly bool _isOrdered;
+            private int _keyClashFoundIteration = 0;
+
+            public PSDictionary(bool isOrdered) {
+                _isOrdered = isOrdered;
+
+                // By default use a non cases-sensitive comparer
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/serialization.cs:4002]:
```suggestion
                // By default use a non case-sensitive comparer
```

**Review by SteveL-MSFT — APPROVED** [review]

**Review by TravisEz13 — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
@davidBar-On Thanks for your contribution!

**Pull request merged**